### DISCUSSION TITLE
How (if at all) should Meltano Cloud leverage the Meltano system DB?

### DISCUSSION THREAD

**Discussion by WillDaSilva** [meltano/meltano#7827] [General]
Follows from:
- https://github.com/meltano/infra/issues/1416#issuecomment-1601074219

My initial comment:

> Meltano core already uses mutexes to prevent the same schedule from running multiple times in parallel, but as @pnadolny13 called out:
>
 >>   but I think we've implemented cloud without a standard meltano system db.
>
> I think that post-GA we'll have to decide on and implement some way for Meltano Cloud to use the Meltano system DB in a persistent way.
>
> We'll probably want to make those system DBs project-specific, which will impose some limitations. For instance, if you have a deployment for an old version of your project, and it runs schedule XYZ, then while that old version of the project is running schedule XYZ you won't be able to run a schedule of the same name in a new deployment with the current version of the project.
>
> I worry that if we don't start using the system DB as intended, the behaviour of Meltano core and Meltano Cloud will drift apart, as demonstrated by this bug.

@tayloramurphy's response:

> At first glance I think it makes sense to try to maintain parity between Cloud and Core. If folks come to Cloud with certain expectations from Core, then it's going to be a bad experience to get onto Cloud. I don't have a good mental model of what having a system db for Cloud actually looks like, but definitely worth discussing more.

@pnadolny13's response:

> I wasn't part of any real design discussions related to whether we should use system dbs in cloud or not but I do remember hearing AJ talk about being concerned with OLTP databases not scaling for cloud and the cost of having a db per user. Its probably worth resisting that discussion but personally it feels like the wrong move to revert to a system db for cloud.
>
> It felt like for a while there we were trending in a direction where meltano core would no longer need a system db anyways (i.e. the movement towards state backends). They're a nuisance for most self hosted users too because they barely hold any data and are usually one of the more expensive pieces of a deployment.

---

Questions to discuss:
- Should Meltano Cloud try to maintain parity with Meltano Core - particularly around the system DB? How important is it for us to maintain parity, and what are the consequences of not doing so? What opportunities are we missing by prioritizing maintaining parity?
- Should Meltano Cloud make use of a persistent system DB?
- Should Meltano Core make use of a persistent system DB?
    - How realistic would it be for us to deprecate the system DB, i.e. make it unnecessary and not used by default for new projects, but not necessarily remove it?
    - What sort of timeline would be realistic for this deprecation?
- If Meltano Cloud uses a persistent system DB, how will it be maintained/upgraded? Should customers have any direct access/control over it? Note that "bring you own system DB" is already supported (as far as I know) by setting the system DB env var.

**Comment by pnadolny13** [comment]
From our office hours discussion and https://docs.meltano.com/concepts/project#system-database it sounds like the system db has these pieces of functionality:

1. State bookmark tracking
2. Tracking job history
3. Locking to disallow 2 jobs at the same time
4. User management
5. Settings management

Ideally we'd have support for all these features without needing a system db, we can still leave the system db as default but allow users to configure other options.

### Self hosted:
1. Solved by state backends https://docs.meltano.com/concepts/state_backends#supported-backends
2. TODO - will recreate a local sqlite system db if it doesnt exist and uses that
3. TODO - will recreate a local sqlite system db if it doesnt exist and uses that
4. Used for UI which is deprecated
5. Used for UI which is deprecated. This might need more exploring, it seems like secrets backends.

### Cloud:
1. Also solved by state backends. Correct me if I'm wrong and we implemented something custom.
2. Solved by custom tracking of job history.
3. TODO - currently allows multiple instance to run at the same time
4. Used for UI which is deprecated
5. Used for UI which is deprecated. This might need more exploring, it seems like secrets backends.

Did I miss anything?

**Reply by pnadolny13** [reply]
Since we need 3 for core and cloud it makes sense to me to implement it generically in core so both benefit vs a custom cloud specific solution.

I also suggested that since we have a locking mechanism already for state files to make them transactionally safe, it might be possible to use the same mechanism to lock jobs while one is running. Idk the implementation details but could be something to consider.

**Comment by kgpayne** [comment]
@WillDaSilva My recollection of the discussions pre-Cloud about the System DB:

1.  It was generally recognised that an OLTP DB shouldn't be the _only_ option for the functions performed by the System DB currently (Secrets, State/Bookmarks, Run History, Logs and Locking), esp. with Cloud and other distributed/serverless deployment modes in mind (including Lambda and GitHub Actions).
2. A model similar to State backends was assumed, whereby users could configure their desired backend for each function. Local SQLite would be retained as default. As a first pass, we talked about a shortlist of 'blessed' backends to include common requests (e.g. DynamoDB, S3 etc.) rather than necessarily allowing users to 'plug in' their own backends via some additional installable (e.g. via `meltano add utility meltano-locking-backend-mongodb`, where `meltano-locking-backend-mongodb` is an EDK plugin implementing an interface for getting/checking/releasing locks).
3. We also discussed Cloud itself being a backend option, akin to the Pulumi freemium model. I.e. we allow users to create a free Cloud account to store their State, Run History/Logs and Schedule/Pipeline Locks as a means of simplifying the overall user experience, identifying potential customers and smoothing their journey to Cloud (i.e. "you already have a login, state and history - why not schedule and run your jobs in Cloud too?"). In Cloud runners we might configure the Cloud backend if none is provided.

On that basis, I'd love to see us tackle 2. (decoupling System DB functions from an OLTP DB) as a precursor to 3. 🙂
